Make eye composite compliant

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82114

This only becomes visible after
https://github.com/pytorch/pytorch/pull/81984

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Make eye composite compliant

This only becomes visible after
https://github.com/pytorch/pytorch/pull/81984

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

Lezcano commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2192,9 +2192,12 @@
 
 - func: eye.m(int n, int m, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
+  dispatch:
+    CompositeExplicitAutograd: eye
```
Why is this `CompositeExplicit`? I think it doesn't have a derivative, does it?

ezyang replied:
So what's actually going on here is that CompositeImplicitAutograd get decomposed automatically for purposes of Python key tracing after https://github.com/pytorch/pytorch/pull/81984 but CompositeExplicitAutograd does not. But eye.m decomposes into an empty alloc and then an out= call, and I don't want that to happen, I want it to get traced as a single unit. Adding this line forces this behavior.

Lezcano replied:
But wouldn't this have some secondary effects like weirder errors of the form "`eye` does not have a derivative" or so if you try to differentiate through it? Or is it fine because it always returns a tensor with `requires_grad=False`? What about forward AD? CC @albanD

albanD replied:
Actually that is not true, you can do `torch.eye(2, requires_grad=True)`. But the autograd is already handled at a higher level and in a special case for factory functions.

@ezyang I'm not sure to understand why you need to change that though? This value is ignored for factory functions anyways?

zou3519 replied:
- In https://github.com/pytorch/pytorch/pull/81984, we make the Python key decompose on CompositeImplicitAutograd.
- The side effect of this is that at::eye, which is CompositeImplicitAutograd, gets decomposed into at::eye_out
- at::eye_out is problematic because it out-variants in general are not "composite compliant" - e.g. doesn't work for all Tensor Subclasses (vmap, autograd). Technically this usage of eye_out is acceptable but our composite compliance testing catches usages of out= variants in composite operations.
- So the resolution is to make eye CompositeExplicitAutograd, like we have done to e.g. at::empty_like

Lezcano replied:
Note that for `empty_like`, when we did so, we added the following entry in `derivatives.yaml`:
```yaml
- name: empty_like(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor
  self: non_differentiable
  output_differentiability: [False]
```
Should we do the same for `eye` then?

ezyang replied:
I think it is not necessary as there are no tensor inputs

Lezcano commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -416,7 +421,7 @@ Tensor eye(int64_t n,
     c10::optional<bool> pin_memory) {
   // the default value of `m` equals to `n`
-  return native::eye(n, n, dtype, layout, device, pin_memory);
+  return at::_ops::eye_m::call(n, n, dtype, layout, device, pin_memory);
```
What's this `at::_ops::eye_m` notation? First time I see it.

ezyang replied:
Hmm, I thought we don't support unpacked TensorOptions at::eye, but looks like we do. This version is usually called from codegen, but it's codegen friendly API where none of the TensorOptions c++ api packing has happened. Calls like at::eye are internally implemented with it.

ezyang committed:
Update on "Make eye composite compliant"


This only becomes visible after
https://github.com/pytorch/pytorch/pull/81984

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

Chillee reviewed (approved):
Looks fine to me, not sure if @albanD or co have other complaints.

lezcano reviewed (approved):
lgtm

albanD commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -17,6 +17,11 @@
 #include <c10/util/irange.h>
 #include <ATen/NamedTensorUtils.h>
 #include <ATen/native/UnaryOps.h>
+#ifndef AT_PER_OPERATOR_HEADERS
```
Per operator header massively speed up incremental build.
But because of internal BUCK, we need a way to disable them :(

ezyang replied:
It is for the call to `at::eye``

albanD reviewed (approved):
SGTM

I do think the factory function bypassing dispatch in weird ways is a problem that will soon hit us though :/

ezyang commented:
@pytorchbot merge
